### PR TITLE
Add (very) initial support for the metrics infrastructure

### LINKED ISSUE #320: Add macro for collecting metrics
Macro similar to the logging macros (`error!`, `trace!`, etc.) should be added for collecting different metrics.

### PR DESCRIPTION
**Pull request by DarkEld3r**
Closes #320.

### CODE REVIEW

**Comment by DarkEld3r** [comment]
@alekseysidorov Nice idea, but it will require a more elaborate implementation that should be flexible enough to use with different backends. I'm not sure that we ready for that (right now). 😄

File: exonum/src/helpers/metrics.rs [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+// limitations under the License.
+
+//! Utilities for collecting metrics.
+
+use std::time::SystemTime;
+use std::fmt::Arguments;
+
+/// Adds given metric with given value.
+///
+/// Metric name should be in the following format: "module_name.metric_name", where "module_name"
```
Comment (aleksuss) [inline exonum/src/helpers/metrics.rs]: Clippy says you should replace double quotes for ` :-)

File: exonum/src/helpers/metrics.rs [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    if cfg!(feature = "metrics-hadoop") {
+        write_to_hadoop(metric_name, value, &time);
+    }
+
+    if cfg!(feature = "metrics-log") {
+        trace!("{} {} {}", metric_name, value, time);
+    }
+}
+
+fn write_to_hadoop(_metric_name: &str, _value: &Arguments, _time: &str) {
```
Comment (aleksuss) [inline exonum/src/helpers/metrics.rs]: What do you think about adding this attribute `#[cfg(feature = "metrics-hadoop")]` ?

Comment (DarkEld3r) [reply exonum/src/helpers/metrics.rs]: Unfortunately, `cfg!(feature = "...")` (macro, not an attribute) works not the same way as C defines, so code under that `cfg!` will still be compiled and cause a compilation error if a function is marked with `#[cfg(...)]` attribute.

File: CHANGELOG.md [hunk 3]
```diff
@@ -9,6 +9,7 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 
 ### Added
 - Implement `FromStr` and `ToString` traits for public sodium types (#318)
+- Base support for metrics infrastructure.
```
Comment (defuz) [inline CHANGELOG.md]: Please be as specific as possible in a changelog. Something like: "Added new macro `metric!` to collect statistical information."

File: exonum/src/helpers/metrics.rs [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+    #[cfg(feature = "metrics-log")]
+    {
+        trace!("{} {} {}", metric_name, value, time);
+    }
+}
+
+#[cfg(feature = "metrics-hadoop")]
+fn write_to_hadoop(_metric_name: &str, _value: &Arguments, _time: &str) {
```
Comment (defuz) [inline exonum/src/helpers/metrics.rs]: Please comment or completely delete this function entirely.

File: exonum/src/node/consensus.rs [hunk 5]
```diff
@@ -461,6 +461,13 @@ where
 
         let height = self.state.height();
 
+        let mempool_size = self.state
+            .transactions()
+            .read()
+            .expect("Expected read lock")
+            .len();
+        metric!("node.mempool", "{}", mempool_size);
```
Comment (defuz) [inline exonum/src/node/consensus.rs]: Why do we need to format here? Any metric is always just a number. So I prefer to see: 

```rust
metric!("node.mempool", mempool_size);
```

**Pull request merged**